README.md
# Todo

A Clojure project created to understand the flow of a web app developed with clojure. We have used clojure + postgresql primarily in this project and gradually added datomic in place of postgresql.

## Requirements

This project will require the installation of clojure related dependencies and postgresql as database on the system (to run the code from master branch).

We have added a datomic branch on which we are migrating the current project from clojure + postgresql to clojure + datomic.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-01-16
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-01-16
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/todo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/todo/compare/0.1.0...0.1.1


project.clj
(defproject todo "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [ring "1.8.0"]
                 [compojure "1.6.1"]
                 [com.datomic/datomic-free "0.9.5697"]
                 [org.clojure/java.jdbc "0.7.11"]
                 [org.postgresql/postgresql "42.2.9"]
                 [hiccup "1.0.5"]]
  :repl-options {:init-ns todo.core}
  :main todo.core
  :datomic {:schemas ["resources/datomic" ["schema.edn"]]}
  :profiles {:dev
             {:main todo.core/-dev-main
              :datomic {:config "resources/datomic/free-transactor-template.properties"
                        :db-uri "datomic:free://localhost:4334/todo"}}})


src/todo/item/view.clj
(ns todo.item.view
  (:require [hiccup.page :refer [html5]]
            [hiccup.core :refer [html h]]))

(defn new-item []
  (html
   [:form.form-horizontal
    {:method "POST" :action "/items"}
    [:div.form-group
     [:label.control-label.col-sm-2 {:for :name-input}
      "Name"]
     [:div.col-sm-10
      [:input#name-input.form-control
       {:name :name
        :placeholder "Name"}]]]
    [:div.form-group
     [:label.control-label.col-sm-2 {:for :desc-input}
      "Description"]
     [:div.col-sm-10
      [:input#desc-input.form-control
       {:name :description
        :placeholder "Description"}]]]
    [:div.form-group
     [:div.col-sm-offset-2.col-sm-10
      [:input.btn.btn-primary
       {:type :submit
        :value "New Todo"}]]]]))

(defn delete-item-form [id]
  (html
   [:form
    {:method "POST" :action (str "/items/" id)}
    [:input {:type :hidden
             :name "_method"
             :value "DELETE"}]
    [:div.btn-group
     [:input.btn.btn-danger.btn-xs
      {:type :submit
       :value "Delete"}]]]))

(defn update-item-form [id checked]
  (html
   [:form
    {:method "POST"
     :action (str "/items/" id)}
    [:input {:type :hidden
             :name "_method"
             :value "PUT"}]
    [:input {:type :hidden
             :name "checked"
             :value (if checked "false" "true")}]
    [:div.btn-group
     [:button.btn.btn-primary.btn-xs
      (if checked "DONE" "TODO")]]]))

(defn items-page [items]
  (html5 {:lang :en}
         [:head
          [:title "Listronica"]
          [:meta {:name :viewport
                  :content "width=device-width, initial-scale=1.0"}]
          [:link {:href "/bootstrap/css/bootstrap.min.css"
                  :rel :stylesheet}]]
         [:body
          [:div.container
           [:h1 "My Todos"]
           [:div.row
            (if (seq items)
              [:table.table.table-striped
               [:thead
                [:tr
                 [:th "Name"]
                 [:th.col.sm-4 "Description"]
                 [:th.col.sm-2 "Action"]
                 [:th.col.sm-2 "Status"]]]
               [:tbody
                (for [i items]
                  [:tr
                   [:td (h (:name i))]
                   [:td (h (:description i))]
                   [:td (delete-item-form (:id i))]
                   [:td (update-item-form (:id i) (:checked i))]])]]
              [:div.col-sm-offset-1 "There are no Todos."])]
           [:div.col-sm-6
            [:h2 "create a new Todo"]
            (new-item)]]
          [:script {:src "http://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"}]
          [:script {:src "/bootstrap/js/bootstrap.min.js"}]]))

src/todo/item/model.clj
(ns todo.item.model
  (:require [clojure.java.jdbc :as db]))

(defn create-table [db]
  (db/execute!
   db
   ["CREATE EXTENSION IF NOT EXISTS \"uuid-ossp\""])
  (db/execute!
   db
   ["CREATE TABLE IF NOT EXISTS items
 (id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
 name TEXT NOT Null, 
 description TEXT NOT NULL,
 checked BOOLEAN NOT NULL DEFAULT FALSE,
 date_created TIMESTAMPTZ NOT NULL DEFAULT now())"]))

(defn create-item [db name description]
  (:id (first (db/query
               db
               ["INSERT INTO items (name, description) values (?,?)
                RETURNING id"
                name
                description]))))

(defn update-item [db id checked]
  (= [1] (db/execute!
          db
          ["UPDATE items
          SET CHECKED = ?
          WHERE id = ?"
           checked
           id])))

(defn delete-item [db id]
  (= [1] (db/execute!
          db
          ["DELETE FROM items
          WHERE id = ?"
           id])))


(defn read-items [db]
  (db/query
   db
   ["SELECT id, name, description, checked, date_created FROM items
ORDER BY date_created"]))

src/todo/item/handler.clj
(ns todo.item.handler
(:require [todo.item.model :refer [create-item read-items update-item delete-item]]
            [todo.item.view :refer [items-page]]))

(defn handle-index-items [req]
(let [db (:todo/db req)
      items (read-items db)]
      {
          :status 200
          :headers {}
          :body (items-page items)
        ;   (str "<html><head></head><body><div>"
        ;   (mapv :name items)
        ;   "</div><form method =\"POST\" action=\"/items\">"
        ;   "<input type=\"text\" name= \"name\" placeholder=\"name\">"
        ;   "<input type=\"text\" name= \"description\" placeholder=\"description\">"
        ;   "<input type=\"submit\">"
        ;   "</form</body></html>")
      }))

(defn handle-create-item [req]
(let [name (get-in req [:params "name"])
      description (get-in req [:params "description"])
      db (:todo/db req)
      item-id (create-item db name description)]
      {:status 302
      :headers {"Location" "/items"}
      :body ""}))

(defn handle-delete-item [req]
(let [db (:todo/db req)
      item-id (java.util.UUID/fromString (:item-id (:route-params req)))
      exists? (delete-item db item-id)]
      (if exists?
      {:status 302
      :headers {"Location" "/items"}
      :body ""}
      {:status 404
      :headers {}
      :body "Item not found."})))


(defn handle-update-item [req]
(let [db (:todo/db req)
      item-id (java.util.UUID/fromString (:item-id (:route-params req)))
      checked (get-in req [:params "checked"])
      exists? (update-item db item-id (= "true" checked))]
      (if exists?
      {:status 302
      :headers {"Location" "/items"}
      :body ""}
      {:status 404
      :headers {}
      :body "Item not found."})))

src/todo/core.clj
(ns todo.core
(:require [todo.item.model :as items]
          [todo.item.handler :refer [handle-index-items
                                     handle-create-item
                                     handle-delete-item
                                     handle-update-item]]
          [todo.item.handler_datomic :refer [handle-todos-list
                                     handle-create-todo
                                     handle-get-todo]])
(:require [datomic.api :as d])
(:require [ring.adapter.jetty :as jetty]
          [ring.middleware.reload :refer [wrap-reload]]
          [ring.middleware.params :refer [wrap-params]]
          [ring.middleware.resource :refer [wrap-resource]]
          [ring.middleware.file-info :refer [wrap-file-info]]
          [compojure.core :refer [defroutes ANY GET POST PUT DELETE ]]
          [compojure.route :refer [not-found]]
          [ring.handler.dump :refer [handle-dump]]))


(def uri "datomic:free://localhost:4334/todo")

(defn create-empty-in-memory-db []
(let [uri "datomic:free://localhost:4334/todo"]
  (d/delete-database uri)
  (d/create-database uri)
  (let [conn (d/connect uri)
        schema (load-file "resources/datomic/schema.edn")]
    (d/transact conn schema)
    conn)))

(def conn (d/connect uri))
(def db "jdbc:postgresql://localhost/todo?user=postgres&password=postgres")
(create-empty-in-memory-db)

(defn greet [req]

{
  :status 200
  :body "Hello, World"
  :headers {}
})

(defn goodbye [req]
{
  :status 200
  :body "Goodbye, Cruel world!"
  :headers {}
})

(defn about [req]
{
  :status 200
  :body "This is our first clojure application!"
  :headers {}
})

(defn yo [req]
(let [name (get-in req [:route-params :name])]
{
  :status 200
  :body (str "yo! " name "!")
  :headers {}
}))

(def ops
{"+" +
"-" -
"*" *
":" /})

(defn calc [req]
(let [a (Integer. (get-in req [:route-params :a]))
      b (Integer. (get-in req [:route-params :b]))
      op (get-in req [:route-params :op])
      f (get ops op)
]
(if f {
  :status 200
  :body (str (f a b))
  :headers {}
}{
  :status 404
  :body (str "Unknown operator: " op)
  :headers {}
})
))

; (defn request [req]
; {
;   :status 200
;   :body (pr-str req)
;   :headers {}
; })

(defroutes routes
(GET "/" [] handle-index-items)
(GET "/goodbye" [] goodbye)
(GET "/yo/:name" [] yo)
(GET "/about" [] about)
(ANY "/request" [] handle-dump)
(GET "/calc/:a/:op/:b" [] calc)
(GET "/items" [] handle-index-items)
(POST "/items" [] handle-create-item)
(DELETE "/items/:item-id" [] handle-delete-item)
(PUT "/items/:item-id" [] handle-update-item)
(not-found "Page not found.")
(GET "/todo" [] handle-todos-list)
(GET "/todo/:todo-id" [] handle-get-todo)
(POST "/todo" [] handle-create-todo))

(defn wrap-db [hdlr]
(fn [req]
(hdlr (assoc req :todo/db db))))

(defn wrap-server [hdlr]
(fn [req]
(assoc-in(hdlr req)[:headers "Server"] "Listronica 9000")))

(def sim-method {"PUT" :put
                  "DELETE" :delete})

(defn wrap-simulated-methods [hdlr]
(fn [req]
(if-let [method (and (= :post (:request-method req))
(sim-method (get-in req [:params "_method"])))]
(hdlr (assoc req :request-method method))
(hdlr req))))

(def app (wrap-server (wrap-file-info (wrap-resource (wrap-db (wrap-params (wrap-simulated-methods routes))) "static"))))

(defn -main [port]
(items/create-table db)
(jetty/run-jetty app {:port (Integer. port)}))

(defn -dev-main [port]
(items/create-table db)
(jetty/run-jetty (wrap-reload #'app){:port (Integer. port)}))
